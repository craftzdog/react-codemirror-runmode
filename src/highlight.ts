import { Parser } from '@lezer/common'
import {
  HighlightStyle,
  Language,
  LanguageDescription
} from '@codemirror/language'
import { highlightTree } from '@lezer/highlight'
import { languages } from '@codemirror/language-data'

export async function getCodeParser(
  languageName: string,
  defaultLanguage?: Language
): Promise<Parser | null> {
  if (languageName && languages) {
    const found = LanguageDescription.matchLanguageName(
      languages,
      languageName,
      true
    )
    if (found instanceof LanguageDescription) {
      if (!found.support) await found.load()
      return found.support ? found.support.language.parser : null
    } else if (found) return (found as any).parser
  }
  return defaultLanguage ? defaultLanguage.parser : null
}

export async function highlightCode<Output>(
  languageName: string,
  input: string,
  highlightStyle: HighlightStyle,
  callback: (
    text: string,
    style: string | null,
    from: number,
    to: number
  ) => Output
): Promise<Output[]> {
  const parser = await getCodeParser(languageName)
  if (parser) {
    const tree = parser.parse(input)
    const output: Array<Output> = []
    let pos = 0
    highlightTree(tree, highlightStyle, (from, to, classes) => {
      if (from > pos)
        output.push(callback(input.slice(pos, from), null, pos, from))
      output.push(callback(input.slice(from, to), classes, from, to))
      pos = to
    })
    pos != tree.length &&
      output.push(
        callback(input.slice(pos, tree.length), null, pos, tree.length)
      )
    return output
  } else {
    return [callback(input, null, 0, input.length)]
  }
}
