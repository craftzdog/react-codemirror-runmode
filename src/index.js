import React from 'react'
import PropTypes from 'prop-types'

export default class MirrorLight extends React.Component {
  static propTypes = {
    codeMirror: PropTypes.func.isRequired,
    className: PropTypes.string,
    theme: PropTypes.string,
    inline: PropTypes.bool,
    language: PropTypes.string,
    prefix: PropTypes.string,
    value: PropTypes.string.isRequired
  }

  static defaultProps = {
    className: '',
    prefix: 'cm-'
  }

  render() {
    const {
      inline,
      codeMirror,
      value,
      language,
      className,
      prefix,
      theme
    } = this.props
    const elements = []
    let index = 0
    let lastStyle = null
    let tokenBuf = ''
    const pushElement = (token, style) => {
      elements.push(
        <span className={style ? prefix + style : ''} key={++index}>
          {token}
        </span>
      )
    }
    const mode = codeMirror.findModeByName(language)
    codeMirror.runMode(value, mode ? mode.mime : language, (token, style) => {
      if (lastStyle === style) {
        tokenBuf += token
        lastStyle = style
      } else {
        if (tokenBuf) {
          pushElement(tokenBuf, lastStyle)
        }
        tokenBuf = token
        lastStyle = style
      }
    })
    pushElement(tokenBuf, lastStyle)

    const code = (
      <code className={inline ? `inline ${prefix}s-${theme}` : ''}>
        {elements}
      </code>
    )

    return inline ? (
      code
    ) : (
      <pre className={`${className} ${prefix}s-${theme}`}>{code}</pre>
    )
  }
}
