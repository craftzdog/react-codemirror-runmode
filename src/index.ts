export * from './highlight.js'
export * from './react-highlighter.js'
