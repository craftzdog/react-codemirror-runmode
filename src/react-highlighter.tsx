import React, { memo, useEffect, useState } from 'react'
import { highlightCode } from './highlight.js'
import type { Highlighter as LezerHighlighter } from '@lezer/highlight'

export type HighlighterProps = {
  lang: string
  children: string
  theme: LezerHighlighter
}

export const Highlighter = memo<HighlighterProps>((props: HighlighterProps) => {
  const { lang, children: code, theme } = props
  const [highlightedCode, setHighlightedCode] = useState<
    React.ReactNode[] | null
  >(null)

  useEffect(() => {
    highlightCode(
      lang,
      code,
      theme,
      (text: string, style: string | null, from: number) => {
        return (
          <span key={from} className={style || ''}>
            {text}
          </span>
        )
      }
    ).then(setHighlightedCode)
  }, [lang, code, theme])

  return <>{highlightedCode || code}</>
})
