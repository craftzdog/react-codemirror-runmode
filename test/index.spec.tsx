// @vitest-environment jsdom
import { describe, expect, it } from 'vitest'
import { Parser } from '@lezer/common'
import { oneDarkHighlightStyle } from '@codemirror/theme-one-dark'
import { getCodeParser, highlightCode, Highlighter } from '../src'
import { render, screen } from '@testing-library/react'

const sleep = (msec: number) =>
  new Promise(resolve => setTimeout(resolve, msec))

describe('getCodeParser', () => {
  it('loads a JavaScript parser', async () => {
    const parser = await getCodeParser('js')
    expect(parser).toBeInstanceOf(Parser)
  })
  it('loads a Python parser', async () => {
    const parser = await getCodeParser('python')
    expect(parser).toBeInstanceOf(Parser)
  })
  it('returns null for non-existing languages', async () => {
    const parser = await getCodeParser('foobar123')
    expect(parser).toBeNull()
  })
})

describe('highlightCode', () => {
  it('highlights JavaScript code', async () => {
    const highlighted = await highlightCode(
      'js',
      'const x = 123',
      oneDarkHighlightStyle,
      (text, style, from, to) => ({ text, style, from, to })
    )
    expect(highlighted).toHaveLength(7)
    expect(highlighted[0]).toEqual({
      text: 'const',
      style: 'ͼp',
      from: 0,
      to: 5
    })
    expect(highlighted[1]).toEqual({
      text: ' ',
      style: null,
      from: 5,
      to: 6
    })
    expect(highlighted[2]).toEqual({
      text: 'x',
      style: 'ͼt',
      from: 6,
      to: 7
    })
    expect(highlighted[4]).toEqual({
      text: '=',
      style: 'ͼv',
      from: 8,
      to: 9
    })
    expect(highlighted[6]).toEqual({
      text: '123',
      style: 'ͼu',
      from: 10,
      to: 13
    })
  })
})

describe('React Highlighter', () => {
  it('renders highlighted code', async () => {
    const code = 'const x = 123'
    render(
      <code className="mde-preview">
        <Highlighter lang="js" theme={oneDarkHighlightStyle}>
          {code}
        </Highlighter>
      </code>
    )
    await sleep(100)
    const element = screen.getByText('const')
    expect(element).toBeDefined()
    expect(element).toHaveProperty('className', 'ͼp')
    expect(screen.getByText('123')).toHaveProperty('className', 'ͼu')
  })
})
