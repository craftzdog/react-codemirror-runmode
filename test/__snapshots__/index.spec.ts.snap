// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`highlightCode > highlights JavaScript code 1`] = `
[
  {
    "from": 0,
    "style": "ͼp",
    "text": "const",
    "to": 5,
  },
  {
    "from": 5,
    "style": null,
    "text": " ",
    "to": 6,
  },
  {
    "from": 6,
    "style": "ͼt",
    "text": "x",
    "to": 7,
  },
  {
    "from": 7,
    "style": null,
    "text": " ",
    "to": 8,
  },
  {
    "from": 8,
    "style": "ͼv",
    "text": "=",
    "to": 9,
  },
  {
    "from": 9,
    "style": null,
    "text": " ",
    "to": 10,
  },
  {
    "from": 10,
    "style": "ͼu",
    "text": "123",
    "to": 13,
  },
]
`;
